### ISSUE TITLE
Document Hyper-V virtual switch creation

### ISSUE THREAD

**Issue by budhrg** [minishift/minishift#265]
Hi,

As a result of this [comment in issue #244](https://github.com/minishift/minishift/issues/244#issuecomment-271573659), we need to document how Windows users should need to configure hyperv virtual switch manager properly in order to get proper setup of minishift.

In case of above issue #244 , I found that there was already another switch created with wrong internet interface which doesn't had network connectivity.  

I am expecting to have pictorial documentation here.

**Comment by budhrg** [comment]
Any suggestion, where can I add this section?   troubleshooting? and then reference it in README to get more visibility.

cc @hferentschik @LalatenduMohanty @praveenkumar

**Comment by thatdocslady** [comment]
@budhrg from the docs perspective it seems that if this is an issue that users will need to work around before they start using minishift, I would put it in the known issues section of the readme. If it's really long to describe (= lots of text) then maybe troubleshooting + link from README would work, but ideally we'd like to have this info in the relevant location of the workflow where the users would need it.

**Comment by budhrg** [comment]
@thatdocslady , actually, this is one of pre-requisite if user want to start minishift in HyperV hypervisor in Windows.
One has to create `Virtual Switch` but as per my experience and user's experience, I found that most of the users are selecting wrong Network Adapter while creating this switch.

Now, you know why I think it is required to document 😄

**Comment by thatdocslady** [comment]
@budhrg that sounds good of course we should document it, I'm mainly trying to see where we do this. The readme is likely to be stripped of detailed installation and getting started info, but for now I would recommend inserting this step in the appropriate place in the workflow (so after installation but before starting, if I understand correctly?).

**Comment by gbraad** [comment]
@budhrg did you have a preview or branch for this?

**Comment by budhrg** [comment]
@gbraad Don't have branch but working along the line of this article https://technet.microsoft.com/en-us/windows-server-docs/compute/hyper-v/get-started/create-a-virtual-switch-for-hyper-v-virtual-machines

**Comment by thesteve0** [comment]
I agree with @budhrg - this is a necessary pre-step **especially** if they already have docker machine installed. 

Docker machine makes an internal adapter and minishift will grab that one as the network adapter which will make the VM non-functional. 

Really Minishift should make it's own external network switch, named minishift0 or something, and bind to that. 

In absence of that we need to give doc on how to make an external adapter and tell people to switch their minishift to this. This has to be on the "getting started page" otherwise users will be in a failed state. 

For example we should probably tell Windows users to:
add themselves to the Hyper-V Admin group
Do minishift start and expect it to fail
Do minishift stop
Go into the Hyper-V console and attach the VM to a pre-existing external switch or make a new one
Then do minishift again and it should work (I think)

**Comment by gbraad** [comment]
I'd like to see an automated approach. Docker for Windows offers the following: https://docs.docker.com/docker-for-windows/#/network (internal switch). External can fail in the situation where no DHCP is offered... likely for a road warrior developer. Note: on virtualbox we run on an internal network. What should we advise?

**Comment by thesteve0** [comment]
I agree with your assessment and solution, but I am not sure what to advise. I think the best we can do is:
1) look for an active ethernet or wlan on the host.
2) If present make an external vswitch on that active connection
3) If not then make an internal vswitch
4) Document for the user what we did on their behalf so if they need to change things they can

**Comment by gbraad** [comment]
@thesteve0 it seems to me "Internal switch" with NAT and inform the user about this? It is possible with PowerShell to automate this:

```powershell
New-VMSwitch -Name "minishift0" -SwitchType NAT -NATSubnetAddress 192.168.22.0/24
New-NetNat -Name minishift0nat -InternalIPInterfaceAddressPrefix "192.168.22.0/24"
```

**Comment by thesteve0** [comment]
Sorry I have no idea - I would only be able to figure it out using google
searches and trial and error. If you do some up with something I would be
glad to test it. About 12 years ago I used to help administer several labs
of windows machines, since then I have mostly been Linux only. I am using a
Windows laptop a lot now so I can live the life of most of the OpenShift
users but I am definitely not a power windows users. Again, sorry

**Comment by gbraad** [comment]
@hferentschik WDYT? Automate using a "Internal switch", or automate by creating a "External switch" and bind an interface, or let the user configure it ?

Anyways, we need to provide a basic documentation for both approaches if setting up the network and explain the differences.

**Comment by hferentschik** [comment]
>  Automate using a "Internal switch", or automate by creating a "External switch" and bind an interface, or let the user configure it ?

Preference would be internal, but my understanding so far was that external was a requirement. If this is not the case I am all for it, if we can automate it. Also if doing via external switch would mean considerably less effort for us, I would pick external.

**Comment by gbraad** [comment]
> external was a requirement

I wonder what makes it a requirement.

**Comment by gbraad** [comment]
Docker for Windows does the following: they have a virtual switch called DockerNAT (internal-only) setup with NAT.
![internal-only](https://cloud.githubusercontent.com/assets/1894/22699925/d0f94040-ed93-11e6-93a9-412f13af884f.png)

**Comment by gbraad** [comment]
@hferentschik part of the requirement is probably related to the fact that `oc` only works correctly with an IPv4 address. With an External network you might be more likely to get the correct address type.

**Comment by gbraad** [comment]
#418

**Comment by hferentschik** [comment]
Merged via pull request #573
